### PR TITLE
Provisioning Links to launch and prefill app with login data

### PR DESCRIPTION
**Pull request by AndyScherzinger**
corresponding PR for #248

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you do a code review, fixed the server refresh and also fixed a rotation issue down the road? :smiley:

**Comment by AndyScherzinger** [comment]
To be tested, does on focus lost still work with the now different server already checked flag.

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 1]
```diff
@@ -192,17 +191,13 @@
...
-            AccountTypeUtils.getAuthTokenTypeSamlSessionCookie(MainApp.getAccountType());
-
+    private final String BASIC_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypePass(MainApp.getAccountType());
+    private final String OAUTH_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypeAccessToken(MainApp.getAccountType());
+    private final String SAML_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypeSamlSessionCookie(MainApp.getAccountType());
 
     /**
      * {@inheritDoc}
-     *
+     * <p>
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: Please remove

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 2]
```diff
@@ -141,7 +140,9 @@
     private static final String KEY_USERNAME = "USERNAME";
     private static final String KEY_PASSWORD = "PASSWORD";
     private static final String KEY_ASYNC_TASK_IN_PROGRESS = "AUTH_IN_PROGRESS";
-    
+    public static final String PROTOCOLL_SUFFIX = "://";
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: Protocol only with one L

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 3]
```diff
@@ -182,26 +182,22 @@
...
-            AccountTypeUtils.getAuthTokenTypeSamlSessionCookie(MainApp.getAccountType());
-
+    private final String BASIC_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypePass(MainApp.getAccountType());
+    private final String OAUTH_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypeAccessToken(MainApp.getAccountType());
+    private final String SAML_TOKEN_TYPE = AccountTypeUtils.getAuthTokenTypeSamlSessionCookie(MainApp.getAccountType());
 
     /**
      * {@inheritDoc}
-     * 
+     * <p>
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: Please remove

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 4]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
+
+                if (loginUrlInfo.serverAddress != null && !mServerIsChecked) {
+                    onUrlInputFocusLost();
+                }
+            }
+        }
+    }
+
+    /**
+     * parsess a URI string and returns a login data object with the information from the URI string.
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: typo parsess

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 5]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
+     *
+     * @param prefix URI beginning, e.g. cloud://login/
+     * @param dataString the complete URI
+     * @return login data.
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: remove dot

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 6]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
+     * @throws IllegalArgumentException when
+     */
+    public static LoginUrlInfo parseLoginDataUrl(String prefix, String dataString) throws IllegalArgumentException {
+        if (dataString.length() < prefix.length()) {
+            throw new IllegalArgumentException("Invalid login URL detected");
+        }
+        LoginUrlInfo loginUrlInfo = new LoginUrlInfo();
+
+        //format is basically xxx://login/server:xxx&user:xxx&password while all variables are optional
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: space after //

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 7]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
+        String[] values = data.split("&");
+
+        if (values.length < 1 || values.length > 3) {
+            // error illegal number of URL elements detected
+            throw new IllegalArgumentException("Illegal number of login URL elements detected: " + values.length);
+        }
+
+        for (String value : values) {
+            if (value.startsWith("user" + LOGIN_URL_DATA_KEY_VALUE_SEPARATOR)) {
+                loginUrlInfo.username = value.substring("user:".length());
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: key separator is freely configurable, but in 318 it is still ":"

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 8]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
+                throw new IllegalArgumentException("Illegal magic login URL element detected: " + value);
+            }
+        }
+
+        return loginUrlInfo;
     }
 
     private void initAuthTokenType() {
-        mAuthTokenType = 
+        mAuthTokenType =
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: should be in one line

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 9]
```diff
@@ -279,34 +263,87 @@ public void onClick(View v) {
...
-                boolean oAuthRequired = 
-                    (mAccountMgr.getUserData(mAccount, Constants.KEY_SUPPORTS_OAUTH2) != null);
-                boolean samlWebSsoRequired = ( 
-                    mAccountMgr.getUserData(
-                        mAccount, Constants.KEY_SUPPORTS_SAML_WEB_SSO
-                    ) != null
+                boolean oAuthRequired = (mAccountMgr.getUserData(mAccount, Constants.KEY_SUPPORTS_OAUTH2) != null);
+                boolean samlWebSsoRequired = (
+                        mAccountMgr.getUserData
+                                (mAccount, Constants.KEY_SUPPORTS_SAML_WEB_SSO) != null
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: same here?

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 10]
```diff
@@ -472,49 +508,48 @@ public boolean onTouch(View view, MotionEvent event) {
...
         /// step 1 - load and process relevant inputs (resources, intent, savedInstanceState)
         String presetUserName = null;
         boolean isPasswordExposed = false;
         if (savedInstanceState == null) {
             if (mAccount != null) {
                 presetUserName =
-                    com.owncloud.android.lib.common.accounts.AccountUtils.
-                        getUsernameForAccount(mAccount);
+                        com.owncloud.android.lib.common.accounts.AccountUtils.
+                                getUsernameForAccount(mAccount);
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: one line?

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 11]
```diff
@@ -591,14 +626,13 @@ private void updateAuthenticationPreFragmentVisibility () {
...
-     * 
-     * Do NOT save {@link #mNewCapturedUriFromOAuth2Redirection}; it keeps a temporal flag, 
-     * intended to defer the processing of the redirection caught in 
-     * {@link #onNewIntent(Intent)} until {@link #onResume()} 
-     * 
+     * <p>
+     * Do NOT save {@link #mNewCapturedUriFromOAuth2Redirection}; it keeps a temporal flag,
+     * intended to defer the processing of the redirection caught in
+     * {@link #onNewIntent(Intent)} until {@link #onResume()}
+     * <p>
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: < p > needed?

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 12]
```diff
@@ -669,14 +705,14 @@ public void onRestoreInstanceState(Bundle savedInstanceState) {
     }
 
     /**
-     * The redirection triggered by the OAuth authentication server as response to the 
+     * The redirection triggered by the OAuth authentication server as response to the
      * GET AUTHORIZATION request is caught here.
-     * 
-     * To make this possible, this activity needs to be qualified with android:launchMode = 
+     * <p>
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: same here

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 13]
```diff
@@ -887,38 +942,38 @@ private void hidePasswordButton() {
...
      * For basic authorization, a check of existence of the root folder is
      * performed.
-     * 
-     * For OAuth, starts the flow to get an access token; the credentials test 
+     * <p>
+     * For OAuth, starts the flow to get an access token; the credentials test
      * is postponed until it is available.
-     * 
+     * <p>
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: where do these < p > come from...

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 14]
```diff
@@ -1894,63 +1945,64 @@ private void doOnResumeAndBound() {
...
-     * @param handler   Object responsible for catching and recovering HTTP authentication fails.
+     *
+     * @param webView Web view to emebd into the authentication dialog.
+     * @param handler Object responsible for catching and recovering HTTP authentication fails.
      */
     public void createAuthenticationDialog(WebView webView, HttpAuthHandler handler) {
 
         // Show a dialog with the certificate info
-        CredentialsDialogFragment dialog = 
+        CredentialsDialogFragment dialog =
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: one line

File: src/com/owncloud/android/authentication/LoginUrlInfo.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.owncloud.android.authentication;
```
Comment (tobiasKaminsky) [inline src/com/owncloud/android/authentication/LoginUrlInfo.java]: GPL3, nextcloud header, etc, please

**Comment by tobiasKaminsky** [comment]
:+1: :+1: for the tests!

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java (hunk 13 above)
Comment (AndyScherzinger) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: auto formatter I guess, will fix this

File: src/com/owncloud/android/authentication/AuthenticatorActivity.java (hunk 9 above)
Comment (AndyScherzinger) [inline src/com/owncloud/android/authentication/AuthenticatorActivity.java]: This one is actually formatted right ;) - Way more than 120 characters.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky changes due to code review are done :)

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you do another check and approve if you are fine with the code review changes :grin:

**Comment by tobiasKaminsky** [comment]
:+1:

**Pull request merged**